service/storage_service.cc

```diff
@@ -2158,102 +2158,134 @@ future<> storage_service::decommission() {
             }
+            slogger.info("removenode[{}]: Started removenode operation, removing node={}, sync_nodes={}, ingore_nodes={}", uuid, endpoint, nodes, ignore_nodes);
 
-            if (ss._gossiper.get_live_members().contains(endpoint)) {
-                throw std::runtime_error(format("Node {} is alive and owns this ID. Use decommission command to remove it from the ring", endpoint));
-            }
+            // Step 2: Prepare to sync data
+            std::unordered_set<gms::inet_address> nodes_unknown_verb;
+            std::unordered_set<gms::inet_address> nodes_down;
+            auto req = removenode_request{removenode_cmd::prepare, uuid, leaving_nodes, ignore_nodes};
+            try {
+                parallel_for_each(nodes, [&ss, &req, &nodes_unknown_verb, &nodes_down, uuid] (const gms::inet_address& node) {
+                    return ss._messaging.local().send_removenode(netw::msg_addr(node), req).then([uuid, node] (removenode_response resp) {
+                        slogger.info("removenode[{}]: Got prepare response from node={}", uuid, node);
+                    }).handle_exception_type([&nodes_unknown_verb, node, uuid] (seastar::rpc::unknown_verb_error&) {
+                        slogger.warn("removenode[{}]: Node {} does not support removenode verb", uuid, node);
+                        nodes_unknown_verb.emplace(node);
+                    }).handle_exception_type([&nodes_down, node, uuid] (seastar::rpc::closed_error&) {
+                        slogger.warn("removenode[{}]: Node {} is down for removenode_cmd verb", uuid, node);
+                        nodes_down.emplace(node);
+                    });
+                }).get();
+                if (!nodes_unknown_verb.empty()) {
+                    auto msg = format("removenode[{}]: Nodes={} do not support removenode verb. Please upgrade your cluster and run removenode again.", uuid, nodes_unknown_verb);
+                    slogger.warn("{}", msg);
+                    throw std::runtime_error(msg);
+                }
+                if (!nodes_down.empty()) {
+                    auto msg = format("removenode[{}]: Nodes={} needed for removenode operation are down. It is highly recommended to fix the down nodes and try again. To proceed with best-effort mode which might cause data inconsistency, run nodetool removenode --ignore-dead-nodes <list_of_dead_nodes> <host_id>. E.g., nodetool removenode --ignore-dead-nodes 127.0.0.1,127.0.0.2 817e9515-316f-4fe3-aaab-b00d6f12dddd", uuid, nodes_down);
+                    slogger.warn("{}", msg);
+                    throw std::runtime_error(msg);
+                }
 
-            // A leaving endpoint that is dead is already being removed.
-            if (tmptr->is_leaving(endpoint)) {
-                slogger.warn("Node {} is already being removed, continuing removal anyway", endpoint);
-            }
+                // Step 3: Start to sync data
+                req.cmd = removenode_cmd::sync_data;
```

Comment:
What happens if a node that has prepared restarts? What happens if the node that initiates removenode restarts? Does the prepared node stay prepared forever?

Reply:
Yes, it stays forever until it is cancelled. The node initiates removenode is stateless.

Reply:
How can it be cancelled?

We need a better solution. Each thing creates a new set of decisions to be taken by the operator. This is not documented anywhere, and even if it is, we have to teach scylla-operator about it.

Reply:
For example, 

nodetool removenode abort $ops_uuid.

I am thinking about revert-automatically-if-removenode-coordinator-does-not-ping-in-some-time.

Reply:
This has to be run on all nodes? Or just one node and it propagates it to the cluster?

Using a timeout is possible, but also adds complications.

Reply:
Either admin or the nodes that have responded removenode_cmd::prepare have to revert it.

All nodes that have marked the node to be removed in the pending endpoints have to revert it. So the idea is that once a node has responded removenode_cmd::prepare, it starts a watchdog timer. The removenode coordinator has to feed the watdchdog otherwise the effect of removenode_cmd::prepare is reverted. 

The watchdog logic does add extra complexity. We need to store a handler on the nodes (in current patch, we do not need to, it is almost stateless). The reverted pending endpoints will cause the whole cluster to have extra writes. It is best we can revert it automatically. 

Please note, this is not a regression. Current situation is 10x worse.